lib/SIL/SILPrinter.cpp

```diff
@@ -618,7 +622,21 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
     }
     *this << '\n';
 
+    const auto &SM = BB->getModule().getASTContext().SourceMgr;
+    Optional<SILLocation> PrevLoc;
     for (const SILInstruction &I : *BB) {
+      if (SILPrintSourceInfo) {
+        auto CurSourceLoc = I.getLoc().getSourceLoc();
+        if (CurSourceLoc.isValid()) {
+          if (!PrevLoc || SM.getLineNumber(CurSourceLoc) != SM.getLineNumber(PrevLoc->getSourceLoc())) {
```

Comment:
Can you handle the case where the location distance spans multiple lines? Such as a call with arguments on each line?

I'm not sure if we're currently verifying that locations monotonically increase (@dcci ?), but you may want to check that the current location is greater than the previous location before printing anything.

Reply:
I updated with the > check to compare source locations. And with that a call like:
```
@inline(never)
func foo() { 
  bar(1, 
  2,
  3)
}
```
would look like:

```
// foo()
sil hidden [noinline] @$s9printtest3fooyyF : $@convention(thin) () -> () { 
bb0:
  //   bar(1,
  %0 = integer_literal $Builtin.Int64, 1          // user: %1
  %1 = struct $Int (%0 : $Builtin.Int64)          // user: %7
  //   2,
  %2 = integer_literal $Builtin.Int64, 2          // user: %3
  %3 = struct $Int (%2 : $Builtin.Int64)          // user: %7
  //   3)
  %4 = integer_literal $Builtin.Int64, 3          // user: %5
  %5 = struct $Int (%4 : $Builtin.Int64)          // user: %7
  // function_ref bar(_:_:_:)
  %6 = function_ref @$s9printtest3baryySi_S2itF : $@convention(thin) (Int, Int, Int) -> () // user: %7
  %7 = apply %6(%1, %3, %5) : $@convention(thin) (Int, Int, Int) -> ()
  // } 
  %8 = tuple ()                                   // user: %9
  return %8 : $()                                 // id: %9
} // end sil function '$s9printtest3fooyyF'
```

Reply:
I don't see a way where I could print all of the call arguments at once looking at the API's available on SourceLoc.

Reply:
Why can't you print all source lines between prevLoc and currLoc? What happens with
```
func foo(a, b, c) {
  bar(
   a,
   b,
   c)
}
```

Reply:
Yeah, for this it prints only the first line: 
```
bb0(%0 : $Int, %1 : $Int, %2 : $Int):
  // func foo(_ a:Int, _ b:Int, _ c:Int) {
  debug_value %0 : $Int, let, name "a", argno 1   // id: %3
  debug_value %1 : $Int, let, name "b", argno 2   // id: %4
  debug_value %2 : $Int, let, name "c", argno 3   // id: %5
  //   bar(a,
  // function_ref bar(_:_:_:)
  %6 = function_ref @$s9printtest3baryySi_S2itF : $@convention(thin) (Int, Int, Int) -> () // user: %7
  %7 = apply %6(%0, %1, %2) : $@convention(thin) (Int, Int, Int) -> () 
  // }  
  %8 = tuple ()                                   // user: %9
  return %8 : $()                                 // id: %9
} // end sil function '$s9printtest3fooyySi_S2itF'
```

I was worried about hoisting and some other optimizations that can print out too many source lines if I try to print between prevLoc and currLoc. 
I think the same issue can happen with checking curLoc > prevLoc now.

Reply:
I'm envisioning that all source lines would be printed in source order. Do you want them shuffled? I'm not sure that would be readable.

Reply:
That was my original intention. Because it could help associate some SILInstruction with its corresponding source statement. 
Printing from prevLoc to curLoc can print in source order. And maybe developers can resort to passing -emit-verbose-sil along with -sil-print-sourceinfo to get a complete picture in cases like hoisting while debugging. 
I'll post an update.

Reply:
I misunderstood your intention. If it's more useful to see the source at a fine granularity as the SIL jumps around, then you should go back to the `PrevLoc != CurLoc` test, but in that case I think you really need to print line numbers to make sense of it.

Reply:
Right, I added the line numbers as well now.

Reply:
@atrick We currently don't have that verification enabled.

Reply:
@dcci Anyway, the verification is for `-Onone` and it seems that the main use case for this is understanding `-O` output.

Reply:
Yes, I agree.